docs: Refinery Readme improvements (#837)
honeycombio/refinery: README.md

## Purpose

> Refinery is a tail-based sampling proxy and operates at the level of an entire [trace](https://docs.honeycomb.io/concepts/tracing/). Refinery examines whole traces and intelligently applies sampling decisions to each trace. These decisions determine whether to include or discard the trace data in the sampled data forwarded to Honeycomb.

Author: ```suggestion
Refinery is a tail-based sampling proxy and operates at the level of an entire [trace](https://docs.honeycomb.io/concepts/tracing/). Refinery examines whole traces and intelligently applies sampling decisions to each trace. These decisions determine whether to keep or drop the trace data in the sampled data forwarded to Honeycomb.
```

Refinery processes must be able to communicate with each other to concentrate traces on single servers.

Within your application (or other Honeycomb event sources), you would configure the `API Host` to be `http(s)://load-balancer/`. Everything else remains the same, such as API key, dataset name, and so on since all that lives with the originating client.

### Minimum Configuration

Every Refinery instance should have a minimum of:

- a `linux/amd64` or `linux/arm64` operating system
- 2GB RAM for each server used
- Access to 2 cores for each server used

In many cases, Refinery only needs one node.
Refinery clusters usually run best with fewer, larger servers; for large installations, 16 cores and 16GB of RAM per server is not unusual (with appropriate tuning).

> If experiencing a large volume of traffic, you may need to scale out to multiple nodes, and likely need a small Redis cluster to handle scaling.

Reviewer: I am uncomfortable with this phrasing. If you have more than 1 node, you need redis (unless you're really hardcore but I don't want to encourage it). You probably don't need more than one redis, though, unless you're just paranoid.

Author: @kentquirk How about this then? It's the most recent version of the same content in the Honeycomb docs site [as seen here](https://docs.honeycomb.io/manage-data-volume/refinery/set-up/#system-requirements): 
> In many cases, Refinery only needs one node. If experiencing a large volume of traffic, you may need to scale out to multiple nodes, and likely need a small Redis instance to handle scaling.

Author: Refer to suggested change below.

You can install Refinery with the following command, which uses the default values file:

```bash
helm repo add honeycomb https://honeycombio.github.io/helm-charts
helm install refinery honeycomb/refinery
```

Alternatively, supply your own custom values file:

```bash
helm install refinery honeycomb/refinery --values /path/to/refinery-values.yaml
```

where `/path/to/refinery-values.yaml` is the file's path.

## Peer Management

When operating in a cluster, Refinery expects to gather all of the spans in a trace onto a single instance so that it can make a trace decision. Since each span arrives independently, each Refinery instance needs to be able to communicate with all of its peers in order to distribute traces to the correct instance.

This communication can be managed in two ways: via an explicit list of peers in the configuration file, or by using self-registration via a shared Redis cache. Installations should generally prefer to use Redis. Even in large installations, the load on the Redis server is quite light, with each instance only making a few requests per minute. A single Redis instance with fractional CPU is usually sufficient.

## Configuration

> Configuration is controlled by Refinery's two configuration files - `config.yaml` for general configuration and `rules.yaml` for sampling configuration.

Reviewer: I deliberately didn't give them names. I won't stop you, but in the past I think it has caused more confusion than it has helped.

Author: @kentquirk How about this instead?
```suggestion
Configuration is controlled by Refinery's two configuration files, which is generally referred to as `config.yaml` for general configuration and `rules.yaml` for sampling configuration.
```

Determining the number of machines necessary in the cluster is not an exact science, and is best influenced by watching for buffer overruns. But for a rough heuristic, count on a single machine using about 2GB of memory to handle 5,000 incoming events and tracking 500 sub-second traces per second (for each full trace lasting less than a second and an average size of 10 spans per trace).

## Understanding Regular Operation

Refinery emits a number of metrics to give some indication about the health of the process. These metrics should be sent to Honeycomb, typically with Open Telemetry, and can also be exposed to Prometheus. The interesting ones to watch are:

- Sample rates: how many traces are kept / dropped, and what does the sample rate distribution look like?

> - `[incoming|peer]_router_\*`: how many events (no trace info) vs. spans (have trace info) have been accepted, and how many sent on to peers?

Reviewer: ```suggestion
- `[incoming|peer]_router_*`: how many events (no trace info) vs. spans (have trace info) have been accepted, and how many sent on to peers?
```

Reviewer: This was just a glitch.

> - `collect_cache_buffer_overrun`: this should remain zero; a positive value indicates the need to grow the size of the Collector's circular buffer (via configuration `CacheCapacity`).

Reviewer: This isn't the otel Collector, and in fact let's drop the term entirely as it's now confusing.
```suggestion
- `collect_cache_buffer_overrun`: this should remain zero; a positive value indicates the need to grow the size of Refinery's circular trace buffer (via configuration `CacheCapacity`).
```

> - `process_uptime_seconds`: records the uptime of each process; look for unexpected restarts as a key towards memory constraints.

Reviewer: Wondering if we should delete this reference, or at least add a note about StressRelief. Our guidance used to be "if it's crashing, give it more memory" but with StressRelief on it doesn't tend to crash.

Author: Feel free to update with StressRelief!

Reviewer: Gonna save that for [2.3](https://github.com/honeycombio/refinery/issues/929)

### Configuration Query

> When using the standard configuration file formats, such as TOML and YAML, check the loaded configuration by using one of the `/query` endpoints from the command line on a server that can access a Refinery host.

Reviewer: I don't want to refer to TOML. I took it out deliberately. If someone wants TOML they can get it from the query endpoint, but let's not point them to it.

Honestly, this feature is now largely obsolete since we have validation.

Author: ```suggestion
Check the loaded configuration by using one of the `/query` endpoints from the command line on a server that can access a Refinery host.
```

`app/app.go` is the main control point. When its `Start` function ends, the program shuts down. It launches two `Router`s which listen for incoming events.

`route/route.go` listens on the network for incoming traffic. There are two routers running and they handle different types of incoming traffic: events coming from the outside world (the `incoming` router) and events coming from another member of the Refinery cluster (`peer` traffic). Once it gets an event, it decides where it should go next: is this incoming request an event (or batch of events), and if so, does it have a trace ID? Everything that is not an event or an event that does not have a trace ID is immediately handed to `transmission` to be forwarded on to Honeycomb. If it is an event with a trace ID, the router extracts the trace ID and then uses the `sharder` to decide which member of the Refinery cluster should handle this trace. If it's a peer, the event will be forwarded to that peer. If it's us, the event will be transformed into an internal representation and handed to the `collector` to bundle spans into traces.

> `collect/collect.go` the Collector is responsible for bundling spans together into traces and deciding when to send them to Honeycomb or if they should be dropped. The first time a trace ID is seen, the Collector starts a timer. If the root span, which is a span with a trace ID and no parent ID, arrives before the timer expires, then the trace is considered complete. The trace is sent and the timer is canceled. If the timer expires before the root span arrives, the trace will be sent whether or not it is complete. Just before sending, the Collector asks the `sampler` for a sample rate and whether or not to keep the trace. The Collector obeys this sampling decision and records it (the record is applied to any spans that may come in as part of the trace after the decision has been made). After making the sampling decision, if the trace is to be kept, it is passed along to the `transmission` for actual sending.

Reviewer: I guess if you insist, but I don't want to overemphasize this term that dates from 5 years ago and is now confusing.

Author: I am suggesting formatting changes here. If you have issues with the long-existing content here, then this is an opportunity for a re-write  (or even a deletion) outside of this PR.
